### PR TITLE
Teams

### PR DESCRIPTION
**Pull request by phawk**
- Add teams, one team per subdomain
- Use TeamUser model to associate users with a team, both for active users and for invites. This allows many users to join many teams.
- Moved role to TeamUser so a user can have a different role per team

### CODE REVIEW

**Comment by phawk** [comment]
Two important paths still to write tests for:

- Test ensuring archived users cannot sign in
- Test user accepting team invite

File: app/controllers/public_controller.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class PublicController < ApplicationController
+  layout "auth"
+  skip_before_action :ensure_team!
+  skip_before_action :require_login
+
+  def index
```
Comment (phawk) [inline app/controllers/public_controller.rb]: This is just the public site, where subdomain is empty or 'www'.

File: app/models/user.rb [hunk 2]
```diff
@@ -28,20 +24,18 @@ class User < ActiveRecord::Base
...
   # Overide devise finder to lookup by username or email
-  def self.find_for_database_authentication(warden_conditions)
-    conditions = warden_conditions.dup
-    if login = conditions.delete(:login)
-      where(conditions.to_h).where(["username = :value OR email = :value", { value: login }]).first
-    else
-      where(conditions.to_h).first
-    end
-  end
+  # def self.find_for_database_authentication(warden_conditions)
```
Comment (colinyoung87) [inline app/models/user.rb]: can this just be chucked if it was for devise?

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -31,7 +30,7 @@
   end
 
   describe ".active" do
-    it "ignores invited users" do
+    xit "ignores invited users" do
```
Comment (colinyoung87) [inline spec/models/user_spec.rb]: what does `xit` do? that supposed to be there?

**Comment by colinyoung87** [comment]
:+1: think i understood most of it haha, might need you to explain the `acts_as_tenant` thing

File: app/models/user.rb (hunk 2 above)
Comment (phawk) [inline app/models/user.rb]: Yeah just wanted to keep the SQL around incase we implement login with either email or username with clearance

File: spec/models/user_spec.rb (hunk 3 above)
Comment (phawk) [inline spec/models/user_spec.rb]: xit basically marks the test as pending and doesn’t run any of the code. Need to rewrite this test and the code to join on `team_users.status = 'active'`

**Comment by phawk** [comment]
[acts_as_tenant](https://github.com/ErwinM/acts_as_tenant) is just a gem that makes segregating data based on a tenant simpler. In our use case the tenant is the Team model and on every page load we lookup the team via the subdomain and set 'current_team'. 

Adding `acts_as_tenant(:team)` to a model means it will automatically scope lookups and creates on that model everywhere in the site with a team_id. It will set team_id automatically on create and it will auto add `where(team_id: current_team.id)` to all lookups, like a default scope like acts_as_paranoid does.

Hope that helps.

**Pull request merged**